### PR TITLE
Python: more call graph tracing

### PR DESCRIPTION
**Pull request by RasmusWL**
This PR expands on the capabilities of the call graph tracer. It's a pretty big one, so it's hard to summarize concisely what has happened... a LOT. Almost everything changed :smile: 

Sadly, I didn't find time to write tests, or invest much in writing documentation for the Python code. I prioritized getting this to a point where it can be used for evaluating quality of the new call-graph resolution.

Overall it's in a state where it can run on the test suite of real projects, but not through `tox`, so some manual adaptation of commands is required for settings up a new projects (although it's fairly easy) -- see `./projects.json` and the README for details.

### CODE REVIEW

File: python/tools/recorded-call-graph-metrics/README.md [hunk 1]
```diff
@@ -4,14 +4,113 @@ also known as _call graph tracing_.
 
 Execute a python program and for each call being made, record the call and callee. This allows us to compare call graph resolution from static analysis with actual data -- that is, can we statically determine the target of each actual call correctly.
 
-This is still in the early stages, and currently only supports a very minimal working example (to show that this approach might work).
+Using the call graph tracer does incur heave toll on the performance. Expect 10x longer to execute program.
```
Comment (tausbn) [inline python/tools/recorded-call-graph-metrics/README.md:7]:
```suggestion
Using the call graph tracer does incur a heavy toll on the performance. Expect 10x longer to execute the program.
```

File: python/tools/recorded-call-graph-metrics/README.md [hunk 2]
```diff
@@ -4,14 +4,113 @@ also known as _call graph tracing_.
...
+```
+
+To view the results, use
+```
+$ head -n 100 projects/*/Metrics.txt
+```
+
+### Expanding set of projects
+
+It should be fairly straight forward to expand the set of projects. Most projects use `tox` for running their tests against multiple python versions. I didn't look into any kind of integration, but have manually picked out the instructions required to get going.
```
Comment (tausbn) [inline python/tools/recorded-call-graph-metrics/README.md:28]:
```suggestion
It should be fairly straightforward to expand the set of projects. Most projects use `tox` for running their tests against multiple python versions. I didn't look into any kind of integration, but have manually picked out the instructions required to get going.
```

File: python/tools/recorded-call-graph-metrics/ql/lib/BytecodeExpr.qll [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+class XMLBytecodeMakeFunction extends XMLBytecodeExpr {
+  XMLBytecodeMakeFunction() { this.hasName("BytecodeMakeFunction") }
+
+  XMLBytecodeExpr get_qualified_name_data() {
+    result.getParent() = this.getAChild("qualified_name")
+  }
+}
+
+class XMLSomethingInvolvingScaryBytecodeJump extends XMLBytecodeExpr {
```
Comment (tausbn) [inline python/tools/recorded-call-graph-metrics/ql/lib/BytecodeExpr.qll:69]: That name... :rofl:

File: python/tools/recorded-call-graph-metrics/ql/lib/RecordedCalls.qll [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+      strictcount(this.getABuiltinCallee()) = 1
+    )
+    or
+    // Handle case where the same function is called multiple times in one line, for
+    // example `func(); func()`. This only works if:
+    // - all the callees for these calls is the same
+    // - all these calls were recorded
+    //
+    // without this `strictcount`, in the case `func(); func(); func()`, if 1 of the calls
+    // is not recorded, we woulld still mark the other two recorded calls as valid
```
Comment (tausbn) [inline python/tools/recorded-call-graph-metrics/ql/lib/RecordedCalls.qll:182]:
```suggestion
    // is not recorded, we would still mark the other two recorded calls as valid
```

File: python/tools/recorded-call-graph-metrics/ql/lib/RecordedCalls.qll [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+  IdentifiedRecordedCall() {
+    strictcount(this.getACall()) = 1 and
+    (
+      strictcount(this.getAPythonCallee()) = 1
+      or
+      strictcount(this.getABuiltinCallee()) = 1
+    )
+    or
+    // Handle case where the same function is called multiple times in one line, for
+    // example `func(); func()`. This only works if:
```
Comment (tausbn) [inline python/tools/recorded-call-graph-metrics/ql/lib/RecordedCalls.qll:177]: Is this actually a common occurrence? (I guess `foo(bar(),bar())` might be, but that doesn't quite fit your definition, from what I understand.)

File: python/tools/recorded-call-graph-metrics/ql/query/Metrics.ql [hunk 6]
```diff
@@ -0,0 +1,56 @@
+import lib.RecordedCalls
+
+// colum i is just used for sorting
```
Comment (tausbn) [inline python/tools/recorded-call-graph-metrics/ql/query/Metrics.ql:3]:
```suggestion
// column i is just used for sorting
```

File: python/tools/recorded-call-graph-metrics/src/cg_trace/utils.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    """
+    for op in [
+        "__lt__",
+        "__le__",
+        "__gt__",
+        "__ge__",
+    ]:
+        old = getattr(cls, op)
+
+        # Fix loop variable capture (py/loop-variable-capture)
```
Comment (tausbn) [inline python/tools/recorded-call-graph-metrics/src/cg_trace/utils.py:14]: I'm not sure I understand the function of this comment. It looks like a note to self?

**Review by tausbn — CHANGES_REQUESTED** [review]
A few comments and suggestions (mostly concerning documentation), and questions, but otherwise this looks good to me.

File: python/tools/recorded-call-graph-metrics/ql/lib/RecordedCalls.qll (hunk 5 above)
Comment (RasmusWL) [reply python/tools/recorded-call-graph-metrics/ql/lib/RecordedCalls.qll:177]: The `bar` part of `foo(bar(), bar())` should also be handled, as long as `bar()` calls the same function.

File: python/tools/recorded-call-graph-metrics/src/cg_trace/utils.py (hunk 7 above)
Comment (RasmusWL) [reply python/tools/recorded-call-graph-metrics/src/cg_trace/utils.py]: yeah, I'm not sure why that got committed in the first place

**Review by tausbn — APPROVED** [review]

**Pull request merged**